File: task3/task3.py
import random as rnd
import time
def main():
    mod = 1000000007
    #[n, k] = [int(a) for a in input().split()]
    #nums = [int(a) for a in input().split()]
    elapsed_time = 0
    for i in range(5):
        n = 100000
        k = 75000
        nums = [rnd.randint(-1000000, 1000000) for i in range(n)]
        print(f"n={n}; k={k}")
        start = time.perf_counter()
        print(get_res(nums, k) % mod)
        end = time.perf_counter()
        elapsed_time += end-start
        print("\n")
    print(f"Avg time {elapsed_time/5}")
    '''
    print("K={k}".format(k=k))
    print("Result: {result}".format(result=res))
    print("Positive numbers({count}): {pos_nums}".format(pos_nums=pos_nums, count = pos_count))
    print("Negative numbers({count}): {neg_nums}".format(neg_nums=neg_nums, count = neg_count))
    '''

def get_res(nums, k):
    if len(nums) == k:
        res = 1
        for elem in nums:
            res *= elem
        return res
    nums.sort()
    neg_count = 0
    for elem in nums:
        if elem >= 0:
            break
        neg_count += 1
    neg_nums = nums[:neg_count]
    neg_nums.sort(reverse=True)
    pos_nums = nums[neg_count:]
    pos_count = len(pos_nums)
    res = 1

    value = 2*int(neg_count/2)
    while value > k:
        value -= 2
    value += pos_count

    start = time.perf_counter()

    if k%2 and value < k:
        print("True")
        for i in range(len(nums)):
            nums[i] = abs(nums[i])
        nums.sort()
        for i in range(k):
            #print(f"{res} * {nums[i]}")
            res *= nums[i]

        res *= -1
    else:
        print("False")
        neg_index = neg_count - 1
        pos_index = pos_count - 1
        if k % 2:
            res *= pos_nums[pos_index]
            pos_index -= 1
            pos_count -= 1
            k -= 1

        block2 = block3 = 0
        current = prev = time.perf_counter()
        while k > 0:
            if k == 3 and neg_count > 1 and pos_count > 2:
                if neg_count == 2 and pos_count == 2:
                    res *= neg_nums[neg_index] * neg_nums[neg_index-1] * pos_nums[pos_index]
                    break
                if pos_count > 2:
                    mul_1 = pos_nums[pos_index] * pos_nums[pos_index-1] * pos_nums[pos_index-2]
                    mul_2 = pos_nums[pos_index] * neg_nums[neg_index] * neg_nums[neg_index-1]
                    if mul_1 > mul_2:
                        res *= mul_1
                        break
                    else:
                        res *= mul_2
                        break
            elif neg_count > 1 and k > 1:
                neg_mul = neg_nums[neg_index] * neg_nums[neg_index-1]
                if pos_count > 1:
                    pos_mul = pos_nums[pos_index] * pos_nums[pos_index-1]
                    if neg_mul >= pos_mul:
                        res *= neg_mul
                        neg_index -= 2
                        neg_count -= 2
                    else:
                        res *= pos_mul
                        pos_index -= 2
                        pos_count -= 2
                else:
                    res *= neg_mul
                    neg_index -= 2
                    neg_count -= 2
                k -= 2
                current = time.perf_counter()
                block2 += current-prev
                prev = time.perf_counter()

            else:
                res *= pos_nums[pos_index]
                pos_index -= 1
                pos_count -= 1
                k -= 1
                current = time.perf_counter()
                block3 += current-prev
                prev = time.perf_counter()
        print(f"\tSubBlock2 {block2}")
        print(f"\tSubBlock3 {block3}")


    end = time.perf_counter()
    print(f"Block 3: {end-start}")
    return res


if __name__ == '__main__':
    main()

File: task3/task3.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


class Task {
    int size, numbersAmount;
    long long* arr;
public:
    int mod;
    Task(int size, int numbersAmount, long long* arr) {
        this->size = size;
        this->numbersAmount = numbersAmount;
        this->arr = arr;
        mod = 1000000007;
    }

    long long getResult() {
        sort(arr, arr+size);
        int left = 0, right = size - 1;
        int count = 0;
        long long result = 1;

        if (numbersAmount == 1) return arr[right] % mod;
        if (arr[right] < 0 && numbersAmount % 2 == 1) {
            for (int i = right; i >= size - numbersAmount; i--) {
                result = ((result % mod) * (arr[i] % mod)) % mod;
            }
            return result;
        }
        if (numbersAmount % 2 == 1 && arr[right] > 0) {
            result = arr[right--] % mod;
            count++;            
        }
        
        while (count < numbersAmount) { 
            if (numbersAmount - count == 1) {
                result = ((result % mod) * (arr[right] % mod)) % mod;
                break;
            }    
            if (numbersAmount - count == 3) {
                long long mul1 = arr[left] * arr[left+1] * arr[right];
                long long mul2 = arr[right] * arr[right-1] * arr[right-2];
                long long mul = (mul1 > mul2) ? mul1 : mul2;
                result = ((result % mod) * (mul % mod)) % mod;
                break;
            } 
            long long leftMul = arr[left] * arr[left+1];
            long long rightMul = arr[right] * arr[right-1];
            if (leftMul > rightMul) {
                result = ((result % mod) * (leftMul % mod)) % mod;
                left += 2;
            }
            else {
                result = ((result % mod) * (rightMul % mod)) % mod;
                right -= 2;
            }
            count += 2;
        }
        return result;
    }

    void print() {
        cout << "[" << size << ", " << numbersAmount << "]: ";
        for (int i = 0; i < size; i++) {
            cout << arr[i] << ' ';
        }
        cout << endl;
    }

};

int main() {
    int n, k;
    cin >> n >> k;
    long long* arr = new long long[n];
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    Task* task = new Task(n, k, arr);  
    long long res = task->getResult();
    while (res < 0) res += task->mod;
    cout << res << endl;
}